src/lib/Microsoft.Health.Extensions.Fhir.R4/Repository/FhirServiceRepository.cs

```diff
@@ -0,0 +1,129 @@
+            _cancellationTokenSource.CancelAfter(60000);
+
+            CancellationToken token = cancellationToken.Equals(null) ? _cancellationTokenSource.Token : cancellationToken;
+
+            return await _fhirClient.SearchAsync(resourceType, query, count, token).ConfigureAwait(false);
+        }
+
+        public async Task<Bundle> SearchForResourceAsync(
+            string url,
+            CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(url, nameof(url));
+            _cancellationTokenSource.CancelAfter(60000);
+
+            CancellationToken token = cancellationToken.Equals(null) ? _cancellationTokenSource.Token : cancellationToken;
+
+            return await _fhirClient.SearchAsync(url, token).ConfigureAwait(false);
+        }
+
+        public async Task<T> ReadResourceAsync<T>(
+            ResourceType resourceType,
+            string resourceId,
+            CancellationToken cancellationToken = default)
+            where T : Resource
+        {
+            EnsureArg.IsNotNull<ResourceType>(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNullOrWhiteSpace(resourceId, nameof(resourceId));
+            _cancellationTokenSource.CancelAfter(60000);
+
+            CancellationToken token = cancellationToken.Equals(null) ? _cancellationTokenSource.Token : cancellationToken;
+
+            return await _fhirClient.ReadAsync<T>(resourceType, resourceId, token).ConfigureAwait(false);
+        }
+
+        public async Task<T> ReadResourceAsync<T>(
+           string uri,
+           CancellationToken cancellationToken = default)
+            where T : Resource
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(uri, nameof(uri));
+            _cancellationTokenSource.CancelAfter(60000);
+
+            CancellationToken token = cancellationToken.Equals(null) ? _cancellationTokenSource.Token : cancellationToken;
+
+            return await _fhirClient.ReadAsync<T>(uri, token).ConfigureAwait(false);
+        }
+
+        public async Task<T> UpdateResourceAsync<T>(
+           T resource,
+           string ifMatchVersion = null,
+           string provenanceHeader = null,
+           CancellationToken cancellationToken = default)
+            where T : Resource
+        {
+            EnsureArg.IsNotNull(resource, nameof(resource));
+            _cancellationTokenSource.CancelAfter(60000);
+
+            CancellationToken token = cancellationToken.Equals(null) ? _cancellationTokenSource.Token : cancellationToken;
+
+            return await _fhirClient.UpdateAsync(resource, ifMatchVersion, provenanceHeader, token).ConfigureAwait(false);
+        }
+
+        public async IAsyncEnumerable<Bundle> IterateOverAdditionalBundlesAsync(
+            Bundle bundle,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            if (bundle.Link.Count != 0)
```

Comment:
>!= [](http://example.com/codeflow?start=34&length=2)

Unfortunately we can't short cut based on the count.  It is possible to get a continuation token with no entries but still have a next link. In that scenario we keep following the token till the link is empty.

Reply:
Updated logic to follow token until token.NextLink equals null